### PR TITLE
fix-1547-Bug-in-_partial_fit_and_predict_iterative-in-train_evaluator.py_

### LINKED ISSUE #1547: Bug in _partial_fit_and_predict_iterative in train_evaluator.py?
Hi, 

Working with [AutoSklearn2Classifier](https://github.com/automl/auto-sklearn/blob/b2ac331c500ebef7becf372802493a7b235f7cec/autosklearn/experimental/askl2.py#L180), I recently got a strange error that might be due to a bug. This is what I am running:

```python
    n_cores = 10
    memory_limit= psutil.virtual_memory().available * 10**-6 // n_cores
    time_left_for_this_task = 7200

    predictor = AutoSklearn2Classifier(memory_limit=memory_limit,
                                       metric=autosklearn_metrics.roc_auc,
                                       n_jobs=n_cores,
                                       time_left_for_this_task=time_left_for_this_task
                                      )

    predictor.fit(x_train, y_train)
    predictor.automl_.runhistory_.data
```

where `x_train` is a pandas dataframe and `y_train` a list. Even though `fit` runs, after inspecting `runhistory_.data` it is clear that all models are crashing in line 741 of /autosklearn/evaluation/train_evaluator.py. It seems like `self.Y_train` in line 741 should be `self.X_train` instead:

https://github.com/automl/auto-sklearn/blob/b2ac331c500ebef7becf372802493a7b235f7cec/autosklearn/evaluation/train_evaluator.py#L738-L744

I found that a workaround for this issue is providing `x_train` as numpy. Also tried providing `x_train` as a list of lists or `y_train` as pandas series, but neither of them solved the issue. I am using auto-sklearn==0.14.7.

I am wondering if this is the expected behavior or actually a bug. Thanks!

### PR DESCRIPTION
**Pull request by eddiebergman**
Fixes the below bug and also provides the `X_data` with every call to `self._loss`, which was missing in some parts.

---
* Closes #1547

### CODE REVIEW

File: autosklearn/evaluation/train_evaluator.py [hunk 1]
```diff
@@ -381,39 +352,24 @@ def fit_predict_and_loss(self, iterative: bool = False) -> None:
...
                         if iterations[i] == 1:
-                            self.Y_train_targets[train_indices] = (
-                                self.Y_train.iloc[train_indices]
-                                if hasattr(self.Y_train, "iloc")
-                                else self.Y_train[train_indices]
-                            )
-                            self.X_targets[i] = (
-                                self.X_train.iloc[test_indices]
-                                if hasattr(self.X_train, "iloc")
-                                else self.X_train[train_indices]
```
Comment (eddiebergman) [inline autosklearn/evaluation/train_evaluator.py:392]: mark

File: autosklearn/evaluation/train_evaluator.py [hunk 2]
```diff
@@ -767,8 +716,12 @@ def partial_fit_predict_and_loss(self, fold: int, iterative: bool = False) -> No
                 test_indices=test_split,
                 add_model_to_self=True,
             )
-            train_loss = self._loss(self.Y_actual_train, train_pred)
-            loss = self._loss(self.Y_targets[fold], opt_pred)
+
+            # This is my best guess at what the X_data for these should be
```
Comment (mfeurer) [inline autosklearn/evaluation/train_evaluator.py:720]: I'm not 100% sure, but do we need this code actually? I have the impression that the function `partial_fit_and_predict_iterative` does all the work that's done here anyways? Or maybe we should move this under the `else` block? As of now I'm also not 100% sure of what variables should be there...

File: autosklearn/evaluation/train_evaluator.py [hunk 3]
```diff
@@ -856,13 +796,9 @@ def _partial_fit_and_predict_iterative(
...
 
-                train_loss = self._loss(
-                    self.Y_train.iloc[train_indices]
-                    if hasattr(self.Y_train, "iloc")
-                    else self.Y_train[train_indices],
-                    Y_train_pred,
-                )
-                loss = self._loss(self.Y_train[test_indices], Y_optimization_pred)
+                train_loss = self._loss(y, Y_train_pred, X_data=X)
+                loss = self._loss(y_test, Y_optimization_pred, X_data=X)
```
Comment (mfeurer) [inline autosklearn/evaluation/train_evaluator.py:800]: Should `X_data` here be `X_test` instead of `X`?

**Review by mfeurer — COMMENTED** [review]
I really like how this PR cuts down bloat. I have left a few minor comments. Could you also please add a unit test?

File: autosklearn/evaluation/train_evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:392]: Yup it was, you can see it in the issue :) All the repetition could lead to very subtle bugs which I guess were hard to identify

File: autosklearn/evaluation/train_evaluator.py [hunk 4]
```diff
@@ -381,39 +352,24 @@ def fit_predict_and_loss(self, iterative: bool = False) -> None:
...
 
+                        y = select(self.Y_train, train_indices)
                         model.iterative_fit(
-                            Xt_array[i],
-                            self.Y_train.iloc[train_indices]
-                            if hasattr(self.Y_train, "iloc")
-                            else self.Y_train[train_indices],
-                            n_iter=n_iter,
-                            **fit_params_array[i],
+                            Xt_array[i], y, n_iter=n_iter, **fit_params_array[i]
```
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:372]: I don't know to be honest, there's a lot of indices and setting of `self.<thing>` randomly throughout the place so it's hard to make sense of, I didn't read the full code, I just tried to keep the behavior the same.

I think the evaluations are a good place to do some refactoring to make it more readable and cut out a few hundred lines of what seems like a lot of repetition. However I would leave this for a seperate PR.

File: autosklearn/evaluation/train_evaluator.py (hunk 2 above)
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:720]: I would leave that up to you, I need to read and understand the whole evaluation module to make an informed decision about it.

File: autosklearn/evaluation/train_evaluator.py (hunk 3 above)
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:800]: Yup, that seems like it should, nice spot!

**Comment by eddiebergman** [comment]
I'm not really sure what to unit test here to be honest, could you be a bit more precise?

File: autosklearn/evaluation/train_evaluator.py (hunk 2 above)
Comment (mfeurer) [reply autosklearn/evaluation/train_evaluator.py:720]: Then I guess we can leave this as it is, but I think we don't need `X_fold` here and can use `self.X_targets[fold]` instead?

**Comment by mfeurer** [comment]
After thinking more about what this PR fixes (making the indexing of pandas dataframes and numpy arrays consistent) we don't need a unit test right now to ensure this - we don't have unit tests for pandas and therefore haven't previously caught this error, and this PR simplifies how we access pandas and we can see from the changelog that it is a clear improvement that right now does not require testing.

**Comment by mfeurer** [comment]
There are also unit tests failing wrt the changes (they are hidden among the other failiing unit tests, so you could maybe rebase on development where they are fixed?)?

File: autosklearn/evaluation/train_evaluator.py (hunk 2 above)
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:723]: I think this was the issue causing test failures!

**Comment by eddiebergman** [comment]
Seems the meta learning timeouts have struck again ... I really don't know why, going to rerun the tests

File: autosklearn/evaluation/train_evaluator.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (eddiebergman) [reply autosklearn/evaluation/train_evaluator.py:395]: So this was effecting the doc tests. Our `X_targets[i]` was a `numpy` array and so it was incorrectly doing `self.X_targets[i] = self.X_train[train_indices]`. However, incorrectly changing this to `test_indices`, as it was before, did not seem to fix the problem.

**Pull request merged**